### PR TITLE
Added addition error logging to graphics.log when SDL fails

### PR DESCRIPTION
**Pull request by Mailaender**
to debug https://github.com/OpenRA/OpenRA/pull/19897.

### CODE REVIEW

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 1]
```diff
@@ -515,13 +515,16 @@ static bool CanCreateGLWindow(GLProfile profile)
 			// Implementation inspired by TestIndividualGLVersion from Veldrid
 
 			// Need to create and destroy its own SDL contexts as a workaround for specific buggy drivers
-			SDL.SDL_Init(SDL.SDL_INIT_VIDEO);
+			if (SDL.SDL_Init(SDL.SDL_INIT_VIDEO) != 0)
+				Log.Write("graphics", "SDL initialisation failed");
```
Comment (abcdefg30) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: We should probably `SDL.SDL_Quit();` here too, shouldn't we?

Comment (pchote) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: https://wiki.libsdl.org/SDL_Quit

It certainly wouldn't do any harm, but the official example doesn't use it.

Comment (Mailaender) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: I wouldn't as the actual error message is logged below and might have additional intel.

Comment (RoosterDragon) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Would be worth logging `SDL.SDL_GetError()` here, since I imagine `SDL_CreateWindow` would overwrite it later.

Comment (Mailaender) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Added it.

**Comment by pchote** [comment]
I'm sorry to say that there is a fundamental issue here - the entire purpose of this method is that we expect some of these calls to fail! This causes unnecessary log spam, for example on macOS I now always see

```
SDL window creation failed: Could not initialize OpenGL / GLES library
SDL window creation failed: Could not initialize OpenGL / GLES library
```

because `GLProfile.ANGLE` and `GLProfile.Embedded` aren't supported on macOS.

We only have a problem here if *all* the profiles fail!

**Comment by Mailaender** [comment]
Same thing on Linux. I reworked things.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 2]
```diff
@@ -156,16 +156,17 @@ public GLProfile[] SupportedGLProfiles
 					.ToArray();
 
 				if (!supportedProfiles.Any())
-					throw new InvalidOperationException("No supported OpenGL profiles were found.");
+					throw new InvalidOperationException($"No supported OpenGL profiles were found: {SDL.SDL_GetError()}");
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: The error reported here is likely to be misleading - it will be from whatever profile was tested last, which is most likely not the profile that we want to be using anyway.

If we want to do this properly we really should be storing all the errors in e.g. a list, mentioning which profile they are associated with, and then report them all if (and only if) there are no supported profiles.

It is very rare that we hit a situation where this information is useful so IMO its not worth the extra complexity.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 3]
```diff
@@ -522,15 +524,13 @@ static bool CanCreateGLWindow(GLProfile profile)
 			var window = SDL.SDL_CreateWindow("", 0, 0, 1, 1, flags);
 			if (window == IntPtr.Zero || !string.IsNullOrEmpty(SDL.SDL_GetError()))
 			{
-				SDL.SDL_ClearError();
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:525]: I'm almost certain that these were intentional, to stop error cases from leaking between profiles.

**Comment by Mailaender** [comment]
I guess the error was detected and we don't need this anymore then.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 4]
```diff
@@ -156,16 +156,17 @@ public GLProfile[] SupportedGLProfiles
...
-					throw new InvalidOperationException("No supported OpenGL profiles were found.");
+					throw new InvalidOperationException($"No supported OpenGL profiles were found: {SDL.SDL_GetError()}");
 
 				profile = supportedProfiles.Contains(requestProfile) ? requestProfile : supportedProfiles.First();
 
 				// Note: This must be called after the CanCreateGLWindow checks above,
 				// which needs to create and destroy its own SDL contexts as a workaround for specific buggy drivers
-				SDL.SDL_Init(SDL.SDL_INIT_VIDEO);
-				SetSDLAttributes(profile);
+				if (SDL.SDL_Init(SDL.SDL_INIT_VIDEO) != 0)
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:171]: I thought that this and the string formatting changes were worthwhile, I just wanted the change above unstaged!

Comment (Mailaender) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:171]: I see. However, the original error was only visible during profile detection, so removing that made everything futile. I added the per profile SDL error log now.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 5]
```diff
@@ -522,6 +529,7 @@ static bool CanCreateGLWindow(GLProfile profile)
 			var window = SDL.SDL_CreateWindow("", 0, 0, 1, 1, flags);
 			if (window == IntPtr.Zero || !string.IsNullOrEmpty(SDL.SDL_GetError()))
 			{
+				errorLog.Add(profile, SDL.SDL_GetError());
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Shouldn't we be logging from `SDL_Init` too? We could use a standard list and prefix the error message with ([profile]) as a string.

Comment (Mailaender) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: I don't really have a test case for `SDL_INIT`. That was mostly inspired by a StyleCop warning.

Comment (pchote) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: This leaves an edge case where we can have no supported profiles but no error message.

Comment (Mailaender) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Simplified the logging and added `SDL_INIT` during profile detection.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 6]
```diff
@@ -510,18 +517,21 @@ static void SetSDLAttributes(GLProfile profile)
 			}
 		}
 
-		static bool CanCreateGLWindow(GLProfile profile)
+		bool CanCreateGLWindow(GLProfile profile)
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: IMO it would be cleaner to keep this static and pass the error list in as an argument. That can then be defined as a local variable within the ctor.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 7]
```diff
@@ -510,18 +517,21 @@ static void SetSDLAttributes(GLProfile profile)
...
 		{
 			// Implementation inspired by TestIndividualGLVersion from Veldrid
 
 			// Need to create and destroy its own SDL contexts as a workaround for specific buggy drivers
-			SDL.SDL_Init(SDL.SDL_INIT_VIDEO);
+			if (SDL.SDL_Init(SDL.SDL_INIT_VIDEO) != 0)
+				errorLog.Add($"SDL init: {SDL.SDL_GetError()}");
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: For consistency we should include the profile here too.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 8]
```diff
@@ -510,18 +517,21 @@ static void SetSDLAttributes(GLProfile profile)
...
 
-		static bool CanCreateGLWindow(GLProfile profile)
+		static bool CanCreateGLWindow(GLProfile profile, List<string> errorLog)
 		{
 			// Implementation inspired by TestIndividualGLVersion from Veldrid
 
 			// Need to create and destroy its own SDL contexts as a workaround for specific buggy drivers
-			SDL.SDL_Init(SDL.SDL_INIT_VIDEO);
+			if (SDL.SDL_Init(SDL.SDL_INIT_VIDEO) != 0)
+				errorLog.Add($"{profile}: SDL init failed: {SDL.SDL_GetError()}");
```
Comment (RoosterDragon) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:528]: If Init fails, should we clear error and return here? It doesn't seem sensible to try and create a window after that.

Comment (abcdefg30) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:528]: https://github.com/OpenRA/OpenRA/pull/19977#discussion_r849630179

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 9]
```diff
@@ -510,18 +517,21 @@ static void SetSDLAttributes(GLProfile profile)
...
 			SetSDLAttributes(profile);
 
 			var flags = SDL.SDL_WindowFlags.SDL_WINDOW_HIDDEN | SDL.SDL_WindowFlags.SDL_WINDOW_OPENGL;
 			var window = SDL.SDL_CreateWindow("", 0, 0, 1, 1, flags);
 			if (window == IntPtr.Zero || !string.IsNullOrEmpty(SDL.SDL_GetError()))
 			{
+				errorLog.Add($"{profile}: {SDL.SDL_GetError()}");
 				SDL.SDL_ClearError();
 				SDL.SDL_Quit();
 				return false;
```
Comment (RoosterDragon) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:541]: We should also log the result of `SDL_GetError` if `SDL_GL_MakeCurrent` fails a few lines below.

Also if the `Distinguish between ANGLE and native GLES` part fails, adding a log message would help indicate that was indeed the failing condition.

**Comment by Mailaender** [comment]
Added more logging.

**Review by RoosterDragon — DISMISSED** [review]

**Comment by anvilvapre** [comment]
#19028

**Review by abcdefg30 — APPROVED** [review]
